### ISSUE TITLE
`Status.Conditions` not being cleared inside AdoptionReconciler

### ISSUE THREAD

**Issue by vijtrip2** [aws-controllers-k8s/community#1168]
ACK Reconciler inside `reconciler.go` clears all the _Status.Conditions_ at the start of reconciliation loop and calculates those _Conditions_ again. This is the correct approach but missing from `adoption_reconciler.go`.

To fix this _AdoptionReconciler_ should also clear all _Conditions_ at the beginning of adoption reconciliation loop and if there was any reconciliation logic dependency on existing Conditions(ex: ACK.Adopted), find another way to execute that logic. Ex: Annotations or Checking adopted Custom Resource etc etc

**Comment by jaypipes** [comment]
I don't quite understand what the problem is here. Please update the title of this issue to reflect what the actual issue is.

If I understand correctly, what SageMaker team is looking for is the *historical record of a resource being adopted by ACK*. As mentioned elsewhere, `Status.Conditions` is not a historical record of all state transitions for a resource. Instead, `Status.Conditions` is the set of conditions that were observed by the controller in the *last reconciliation run* (which is the reason why the `Status.Conditions` is cleared at the beginning of each reconciliation loop).

**Comment by vijtrip2** [comment]
> Status.Conditions is the set of conditions that were observed by the controller in the last reconciliation run (which is the reason why the Status.Conditions is cleared at the beginning of each reconciliation loop).

Exactly and this is missing from AdoptionReconciler. AdoptionReconciler does not clear the conditions at beginning of reconciler loop.

**Comment by surajkota** [comment]
Since the conditions will be cleared on every reconcile loop in future, what is the new mechanism for a user to know if the resource adoption is complete? and the `AdoptedResource` CR can be cleaned up

(not visible to user) adopted resource reconciler also needs a new implementation to determine if this reconcile loop is a no-op or it needs to create the target resource

I propose to introduce a field under status.

**Comment by jaypipes** [comment]
> Since the conditions will be cleared on every reconcile loop in future, what is the new mechanism for a user to know if the resource adoption is complete? and the `AdoptedResource` CR can be cleaned up

@surajkota the mechanism is actually the same as it used to be: the custom resource exists with the referred-to AWS identifier. @RedbackThomson do you think we should add something to the adopted resource reconciler that deletes the AdoptedResource CR when it successfully creates the target "real" CR?

> (not visible to user) adopted resource reconciler also needs a new implementation to determine if this reconcile loop is a no-op or it needs to create the target resource
> 
> I propose to introduce a field under status.

@surajkota you mean a new field in the *AdoptedResource*'s Status, right?

**Comment by surajkota** [comment]
> you mean a new field in the AdoptedResource's Status, right?

yes

> the mechanism is actually the same as it used to be: the custom resource exists with the referred-to AWS identifier.

previously, we relied on `ACK.Adopted` condition to be set to `true` before checking if target CR exists. I think this mechanism is more user friendly because AdoptedResource reconciler would show an error if user specified a conflicting name for target CR(i.e. a CR with same name already exists). We can change it to some other field under AdoptedResource's Status

**Comment by RedbackThomson** [comment]
> previously, we relied on `ACK.Adopted` condition to be set to `true` before checking if target CR exists

I think this is still the way to go. We can calculate this on every reconciliation loop by searching for the target CR and checking that it has the adoption annotation (so that we know it was not created manually). 

> @RedbackThomson do you think we should add something to the adopted resource reconciler that deletes the AdoptedResource CR when it successfully creates the target "real" CR?

I don't see a whole lot of value to it? AdoptedResource acts as the declaration that our controller should adopted an existing resource. It's status shows whether we are in that state (`ACK.Adopted = true`) or whether there was an error achieving that (`ACK.Adopted = true && ACK.Recoverable = <something>`). If the customer doesn't want something to be adopted anymore, it's up to them to remove the declaration?

I do understand, though, that adoption will probably be used imperatively ie. a customer will adopt a resource manually, once, after creating the cluster. So maybe there is value if we view it like that?

**Comment by rahtr** [comment]
> do understand, though, that adoption will probably be used imperatively ie. a customer will adopt a resource manually, once, after creating the cluster. So maybe there is value if we view it like that?

Is there any advantage of keeping the `AdoptedResource` object? Cleaning it automatically looks like a good solution. Right now we are migration 1000s of resourcing taking benefit of the AdoptedResource, however we have to build a solution to delete the AdoptedResource CR once the `ACK.Adopted = true`

**Comment by A-Hilaly** [comment]
>Is there any advantage of keeping the AdoptedResource object?

@rahtr Not that I can think of :). IMO, if it doesn't impact your workload then there is no need to keep it in the cluster... maybe only for historical purposes?

**Comment by rahtr** [comment]
@A-Hilaly exactly. To avoid putting pressure on the API server, I believe the best is to give an ability to the user to delete the AdoptedResource CR automatically once the resource has been adopted successfully. Maybe a flag like ‘deleteOnSuccess’ in the CR itself.
